d2/src/main/java/com/linkedin/d2/balancer/simple/SimpleLoadBalancerState.java

```diff
@@ -627,6 +660,43 @@ public void setDelayedExecution(long delayedExecution)
     _delayedExecution = delayedExecution;
   }
 
+  @Override
+  public Map<URI, TrackerClient> getClientsSubset(String serviceName,
+                                                  ServiceProperties serviceProperties,
+                                                  int partitionId,
+                                                  Map<URI, TrackerClient> potentialClients)
+  {
+    SubsettingStrategy<URI> subsettingStrategy = _subsettingStrategyFactory.get(serviceName, serviceProperties, partitionId);
+    if (subsettingStrategy != null)
+    {
+      Map<URI, Double> pointsMap = potentialClients.entrySet().stream()
+          .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().getPartitionWeight(partitionId)));
+      Map<URI, Double> subsetMap = subsettingStrategy.getWeightedSubset(pointsMap);
+
+      if (subsetMap == null)
+      {
+        return potentialClients;
+      }
+      else
+      {
+        Map<URI, TrackerClient> subsetClients = new HashMap<>();
+        for (Map.Entry<URI, Double> entry: subsetMap.entrySet())
+        {
+          URI uri = entry.getKey();
+          TrackerClient client = potentialClients.get(uri);
+          client.setSubsetWeight(partitionId, subsetMap.get(uri));
```

Comment:
Fixed. I store a subsetMapCache in the loadBalancerState. When a client is querying the subset, it will compare the serviceName, partitionId, and clusterGenerationId against the cache. So we can avoid setting subset weight again and again for each request.